### PR TITLE
Implement Giving sample integration

### PR DESCRIPTION
**Pull request by gcatanese**
Add Giving example

### CODE REVIEW

File: giving-example/public/adyenGiving.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      "Content-Type": "application/json",
+    },
+  });
+
+  return await res.json();
+}
+
+async function handleDonation(donationToken, pspReference, amount) {
+  try {
+    const res = await callServer(`/api/donations?donationToken=${encodeURIComponent(donationToken)}&pspReference=${pspReference}`, amount);
```
Comment (Kwok-he-Chu) [inline giving-example/public/adyenGiving.js:17]: From security point-of-view, is there any reason why we don't add both the donationToken and the pspReference (==originalPspReference from the payment) in the POST payload (similar to amount). Now this is exposed in the URL.

Should the donationToken be kept a secret?

File: giving-example/public/adyenGiving.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      message: "By donating you agree to the %#terms%#",
+      linkText: "terms and conditions",
+      link: "https://www.adyen.com/legal/terms-and-conditions" // Replace with yours
+
+
+    },
+    onDonate: (state, component) => {
+      if(state.isValid) {
+        console.log("Initiating donation");
+        let donationToken = sessionStorage.getItem("donationToken");
```
Comment (Kwok-he-Chu) [inline giving-example/public/adyenGiving.js:65]: We store the plain-text donationToken on the frontend, what are the best practices here. Wondering whether we could have the backend store this value (and do a look-up based on the PspReference)

**Review by Kwok-he-Chu — COMMENTED** [review]
Nice work. I've added a couple of questions @jlengrand @gcatanese

File: giving-example/index.js [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    // invalid hmac: do not send [accepted] response
+    console.log("Invalid HMAC signature: " + notification);
+    res.status(401).send('Invalid HMAC signature');
+  }
+
+});
+
+// Process incoming Giving Webhook: get NotificationRequestItem,
+// consume the event asynchronously, send response ["accepted"]
+app.post("/api/webhooks/giving", async (req, res) => {
```
Comment (Kwok-he-Chu) [inline giving-example/index.js:281]: Any reason why we have to separate these two endpoints? (And why we're not validating the HMAC of this webhook?)

File: giving-example/index.js [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+  try {
+    // unique ref for the transaction
+    const orderRef = uuid();
+    // Allows for gitpod support
+    const localhost = req.get('host');
+    // const isHttps = req.connection.encrypted;
+    const protocol = req.socket.encrypted? 'https' : 'http';    
+    // Ideally the data passed here should be computed based on business logic
+    const response = await checkout.PaymentsApi.payments({
+      amount: { currency, value: 1000 }, // value is 10€ in minor units
```
Comment (Kwok-he-Chu) [inline giving-example/index.js:109]: Amount should be 10000 here I think

File: giving-example/index.js [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+// process payload asynchronously
+function consumeEvent(notification) {
+  // add item to DB, queue or different thread
+  
+}
+
+
+/* ################# UTILS ###################### */
+
+function findCurrency(type) {
```
[the remark this answers is not in the document]
Comment (gcatanese) [reply giving-example/index.js:314]: In `/api/initiatePayment`

File: giving-example/index.js [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+      browserInfo: req.body.browserInfo, // required for 3ds2
+      shopperIP, // required by some issuers for 3ds2
+      returnUrl: `${protocol}://${localhost}/api/handleShopperRedirect?orderRef=${orderRef}`, // required for 3ds2 redirect flow
+      // special handling for boleto
+      paymentMethod: req.body.paymentMethod.type.includes("boleto")
+        ? {
+            type: "boletobancario_santander",
+          }
+        : req.body.paymentMethod,
+      // Below fields are required for Boleto:
```
[the remark this answers is not in the document]
Comment (gcatanese) [reply giving-example/index.js:127]: Good catch, removed Klarna too

File: giving-example/index.js [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+      socialSecurityNumber: req.body.socialSecurityNumber,
+      shopperName: req.body.shopperName,
+      billingAddress:
+        typeof req.body.billingAddress === "undefined" || Object.keys(req.body.billingAddress).length === 0
+          ? null
+          : req.body.billingAddress,
+      deliveryDate: new Date("2017-07-17T13:42:40.428+01:00"),
```
[the remark this answers is not in the document]
Comment (gcatanese) [reply giving-example/index.js:134]: It is required by Klarna and it is now removed

File: giving-example/index.js (hunk 3 above)
Comment (gcatanese) [reply giving-example/index.js:281]: 2 different webooks in CA, Giving doesnt support HMAC validation

File: giving-example/public/adyenGiving.js (hunk 2 above)
Comment (gcatanese) [reply giving-example/public/adyenGiving.js:65]: yes, I think we need to refactor this (and change it in the Java sample too. I will rework it.

File: giving-example/views/index.handlebars [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    <h1>Adyen Giving - Donations demo</h1>
+      <div class="giving-view-container">
+      <p>
+        This is a sample designed to demonstrate the <a href="https://docs.adyen.com/online-payments/donations/web-component">Adyen Giving</a> donations workflow.
+      </p>
+      <p>
+        First make a test payment using one of our test card numbers, so you can see the donation screen appear.
+        You can find more test card numbers <a href="https://docs.adyen.com/development-resources/test-cards/test-card-numbers">here</a>.
+      </p>
+      <p>To learn more about client-side integration solutions, check out <a href="https://docs.adyen.com/checkout">Online
```
[the remark this answers is not in the document]
Comment (gcatanese) [reply giving-example/views/index.handlebars:12]: Uh, the link to donations on docs is already provided, I think one should be enough. I have dropped the last sentence as it is not relevant.

[the remark this answers is not in the document]
Comment (Kwok-he-Chu) [reply giving-example/views/index.handlebars]: Let's remove the last line ` To learn more about Adyen...` and the link as well

File: giving-example/index.js [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+  try {
+
+    const localhost = req.get('host');
+    const protocol = req.socket.encrypted? 'https' : 'http';    
+  
+    const response = await checkout.PaymentsApi.donations({
+      amount: { currency: req.body.currency, value: req.body.value },
+      reference: uuid(), 
+      paymentMethod: {
```
Comment (Kwok-he-Chu) [inline giving-example/index.js:61]: @gcatanese I noticed that we need to pass-on the payment method here, is this something that a user can choose from? From the demo, it's not possible to switch payment methods: the backend needs to know what payment method was used, or we have to pass this data from the frontend.

File: giving-example/public/adyenGiving.js (hunk 2 above)
Comment (gcatanese) [reply giving-example/public/adyenGiving.js]: Refactored to store donation token in session

File: giving-example/index.js (hunk 9 above)
Comment (gcatanese) [reply giving-example/index.js:79]: True, payment is hardcoded as the sample supports only Card

File: giving-example/public/adyenGiving.js (hunk 1 above)
Comment (gcatanese) [reply giving-example/public/adyenGiving.js]: Refactored to store donation token in session

**Comment by gcatanese** [comment]
@Kwok-he-Chu I have address the comments, and refactored the way we manage the donation token

File: giving-example/public/adyenGiving.js (hunk 1 above)
Comment (Kwok-he-Chu) [reply giving-example/public/adyenGiving.js]: Nice work

**Review by Kwok-he-Chu — APPROVED** [review]

**Pull request merged**